# Medical Billing Research Project
## 01 — Data Understanding (Section 3.1)

**Institution:** University of the Western Cape
**Module:** BIA 716 Applied Research Project
**IDE:** PyCharm + Jupyter

### Objective
This notebook covers Section 3.1 of the methodology — Data Understanding.
The purpose is to describe the raw dataset exactly as received, before any
cleaning or transformation. No data modification happens in this notebook.

### Dataset
- **File:** PRIVATE PATIENT.xlsx
- **Location:** `../dataset/PRIVATE PATIENT.xlsx`
- **Description:** Medical billing invoices from a private healthcare practice in South Africa

In [5]:
# ============================================================
# CELL 1: IMPORTS AND LOAD RAW DATA
# Section 3.1 - Data Understanding
# ============================================================

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats

# Display settings
pd.set_option('display.max_columns', None)
pd.set_option('display.width', 1000)
pd.set_option('display.float_format', '{:.2f}'.format)

# ── Load raw dataset ──────────────────────────────────────
df_raw = pd.read_excel(
    r'C:\2026\MedicalBillingResearch\dataset\PRIVATE PATIENT.xlsx',
    engine='openpyxl'
)

# ── Basic structure ───────────────────────────────────────
print("=" * 60)
print("DATASET SHAPE")
print("=" * 60)
print(f"Rows    : {df_raw.shape[0]:,}")
print(f"Columns : {df_raw.shape[1]}")

print("\n" + "=" * 60)
print("COLUMN NAMES AND DATA TYPES")
print("=" * 60)
print(df_raw.dtypes)

print("\n" + "=" * 60)
print("FIRST 3 ROWS")
print("=" * 60)
df_raw.head(3)

DATASET SHAPE
Rows    : 50,000
Columns : 32

COLUMN NAMES AND DATA TYPES
Invoice Number                              str
Entity name                                 str
Main Member                                 str
Medical Aid Number                          str
Invoice Date                     datetime64[us]
Capture Date                     datetime64[us]
Earliest Capture Date            datetime64[us]
Latest Capture Date              datetime64[us]
Service Centre                              str
Treating Doctor                             str
Referring Doctor Name                       str
Scheme                                      str
Option                                      str
Administrator                               str
Posted Billing Group                        str
ICD10 String                                str
Primary ICD10                               str
ICD10 Description                           str
Patient Name                                str
Patient DOB    

,Invoice Number,Entity name,Main Member,Medical Aid Number,Invoice Date,Capture Date,Earliest Capture Date,Latest Capture Date,Service Centre,Treating Doctor,Referring Doctor Name,Scheme,Option,Administrator,Posted Billing Group,ICD10 String,Primary ICD10,ICD10 Description,Patient Name,Patient DOB,Gender,Specialty,Debtor Status,POSSIBLE PMB CALCULATED,CONTRACTED/NON-CONTRACTED,Age Bracket,Total Outstanding,Total Medical Aid Outstanding,Total Patient Outstanding,Total Turnover,Total Cashflow,Total Journals
0,00152340/Q,B058E399B433F37D30833C0D8373C87E,87084DFF236E3094254394E75E940A0A,5058F1AF8388633F609CADB75A75DC9D,2020-01-27,2020-01-27,2020-01-27,2020-02-28,HOSPITAL,129208286763220BFBB2FEDED4C90E38,F3145D6BCC367760F38D9FB5FBFDF713,PRIVATE PATIENT,PRIVATE PATIENT ACUTE,PRIVATE,MAX / INSURED: 300%(20),M25.57,M25.57,"Pain in joint, ankle and foot",2FA03D27F47C0E6C8C3E1C9C2523A4BD,NaN,F,(028) Orthopaedic Surgeon,Normal,NON-PMB,NON-CONTRACTED,180,0.00,0.00,0.00,6360.69,-6360.69,NaN
1,R0006356/X,5A7B591CD584B300A47671317A4A4040,9532DD1C3E6F7DEAD4AC92F4FC24BD88,84C6D22CFDF3772E66B07BDE636E0571,2024-09-05,2024-09-06,2024-09-06,2024-11-11,NETCARE GREENACRES ROOMS PE,DE787C03CA7CC324D40F558AF7637CBE,B50339A10E1DE285AC99D4C3990B8693,PRIVATE PATIENT,PRIVATE PATIENT ACUTE,PRIVATE,NaN,NaN,NaN,NaN,D41D8CD98F00B204E9800998ECF8427E,NaN,NaN,Physician,Normal,NON-PMB,NON-CONTRACTED,120,0.00,0.00,0.00,NaN,0.00,NaN
2,R0005554/P,0B18F1B02FBA01F59E475CAC9A54DA30,E3A0EF8DC527CF47F2D282DCD47CFD4D,B4B147BC522828731F1A016BFA72C073,2024-11-29,2024-12-31,2024-12-31,2024-12-31,NaN,542423E3C665D35B3D46E0E291F74DD2,NaN,PRIVATE PATIENT,PRIVATE PATIENT ACUTE,PRIVATE,NaN,NaN,NaN,NaN,D41D8CD98F00B204E9800998ECF8427E,NaN,NaN,022,Handed over ACK,NON-PMB,NON-CONTRACTED,30,0.00,0.00,0.00,NaN,0.00,NaN


In [6]:
# ============================================================
# CELL 2: VARIABLE CLASSIFICATION SUMMARY
# Section 3.1 - Data Understanding
# ============================================================

# Classify variables by type
datetime_cols = df_raw.select_dtypes(include='datetime64').columns.tolist()
numeric_cols  = df_raw.select_dtypes(include='number').columns.tolist()
object_cols   = df_raw.select_dtypes(include='object').columns.tolist()

print("=" * 60)
print("VARIABLE TYPE CLASSIFICATION")
print("=" * 60)
print(f"\nDatetime variables  ({len(datetime_cols)}):")
for col in datetime_cols:
    print(f"   {col}")

print(f"\nNumeric variables   ({len(numeric_cols)}):")
for col in numeric_cols:
    print(f"   {col}")

print(f"\nCategorical/Text variables ({len(object_cols)}):")
for col in object_cols:
    print(f"   {col}")

print(f"\nTotal columns: {len(datetime_cols) + len(numeric_cols) + len(object_cols)}")

# Missing values overview
print("\n" + "=" * 60)
print("MISSING VALUES — RAW DATASET OVERVIEW")
print("=" * 60)
missing = df_raw.isnull().sum()
missing_pct = (missing / len(df_raw) * 100).round(2)
missing_df = pd.DataFrame({
    'Missing Count': missing,
    'Missing %': missing_pct
}).query('`Missing Count` > 0').sort_values('Missing %', ascending=False)

print(missing_df.to_string())
print(f"\nColumns with no missing values: "
      f"{(missing == 0).sum()} out of {df_raw.shape[1]}")

VARIABLE TYPE CLASSIFICATION

Datetime variables  (4):
   Invoice Date
   Capture Date
   Earliest Capture Date
   Latest Capture Date

Numeric variables   (7):
   Age Bracket
   Total Outstanding
   Total Medical Aid Outstanding
   Total Patient Outstanding
   Total Turnover
   Total Cashflow
   Total Journals

Categorical/Text variables (21):
   Invoice Number
   Entity name
   Main Member
   Medical Aid Number
   Service Centre
   Treating Doctor
   Referring Doctor Name
   Scheme
   Option
   Administrator
   Posted Billing Group
   ICD10 String
   Primary ICD10
   ICD10 Description
   Patient Name
   Patient DOB
   Gender
   Specialty
   Debtor Status
   POSSIBLE PMB CALCULATED
   CONTRACTED/NON-CONTRACTED

Total columns: 32

MISSING VALUES — RAW DATASET OVERVIEW
                       Missing Count  Missing %
Total Journals                 46834      93.67
Patient DOB                    21241      42.48
Total Cashflow                  6683      13.37
ICD10 String                 

C:\Users\Admin\AppData\Local\Temp\ipykernel_25972\1896611043.py:9: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  object_cols   = df_raw.select_dtypes(include='object').columns.tolist()


In [7]:
# ============================================================
# CELL 3: DATASET PROFILE SUMMARY
# Section 3.1 - Data Understanding
# This gives us a complete statistical snapshot of the raw
# dataset for documentation in the report (Section 3.1)
# ============================================================

# Fix the deprecation warning from Cell 2 — explicitly
# separate string columns from other object types
str_cols = df_raw.select_dtypes(include='str').columns.tolist()

# ── Unique value counts per column ───────────────────────
# This helps us immediately spot constants (1 unique value)
# and high-cardinality columns (too many unique values)
print("=" * 60)
print("UNIQUE VALUE COUNTS PER COLUMN")
print("=" * 60)
for col in df_raw.columns:
    n_unique = df_raw[col].nunique()
    # Flag constants and high-cardinality columns
    flag = ""
    if n_unique == 1:
        flag = "  ← CONSTANT (same value in every row)"
    elif n_unique == 50000:
        flag = "  ← UNIQUE PER ROW (identifier)"
    print(f"  {col:<40} {n_unique:>6} unique values{flag}")

# ── Summary statistics for numeric columns ────────────────
# This feeds directly into Table 1 of Section 3.2.2
# (Numeric variables descriptive analysis)
print("\n" + "=" * 60)
print("NUMERIC VARIABLES — DESCRIPTIVE STATISTICS")
print("=" * 60)
numeric_summary = df_raw[numeric_cols].describe().T
numeric_summary['skewness'] = df_raw[numeric_cols].skew()
print(numeric_summary[['count', 'min', 'mean', '50%', 'max', 'std', 'skewness']]\
      .rename(columns={'50%': 'median'})\
      .to_string())

# ── Check the two columns we know are suspicious ─────────
# Scheme and Administrator appeared constant in R — verify here
print("\n" + "=" * 60)
print("CONSTANT COLUMN CHECK — Scheme and Administrator")
print("=" * 60)
print("Scheme unique values      :", df_raw['Scheme'].unique())
print("Administrator unique values:", df_raw['Administrator'].unique())

UNIQUE VALUE COUNTS PER COLUMN
  Invoice Number                            34802 unique values
  Entity name                                 397 unique values
  Main Member                               22084 unique values
  Medical Aid Number                         1584 unique values
  Invoice Date                               1998 unique values
  Capture Date                               1577 unique values
  Earliest Capture Date                      1676 unique values
  Latest Capture Date                        1481 unique values
  Service Centre                              730 unique values
  Treating Doctor                             566 unique values
  Referring Doctor Name                      3741 unique values
  Scheme                                        1 unique values  ← CONSTANT (same value in every row)
  Option                                        6 unique values
  Administrator                                 1 unique values  ← CONSTANT (same value in every ro

In [8]:
# ============================================================
# CELL 4: DEEPER INVESTIGATION OF KEY FLAGS
# Section 3.1 - Data Understanding
# We dig into the suspicious findings from Cell 3 before
# we move to data preparation. This is still pure observation
# — no changes to data yet, just understanding what we have.
# ============================================================

# ── 1. DUPLICATE INVOICE NUMBERS ─────────────────────────
# Invoice Number had 34,802 unique values out of 50,000 rows
# meaning ~15,000 rows share an invoice number with another row
# We need to understand whether these are true duplicates
# or the same invoice appearing across multiple billing events
print("=" * 60)
print("1. DUPLICATE INVOICE NUMBERS")
print("=" * 60)
duplicate_invoices = df_raw[df_raw.duplicated(subset='Invoice Number', keep=False)]
print(f"Rows sharing an Invoice Number : {len(duplicate_invoices):,}")
print(f"Unique duplicated numbers       : {duplicate_invoices['Invoice Number'].nunique():,}")
print(f"\nExample — top 5 most repeated Invoice Numbers:")
print(df_raw['Invoice Number'].value_counts().head())

# ── 2. DEBTOR STATUS — ALL 20 CATEGORIES ─────────────────
# We expected ~8-9 categories but found 20
# Listing all of them to spot duplicates, casing issues, typos
print("\n" + "=" * 60)
print("2. DEBTOR STATUS — ALL UNIQUE VALUES")
print("=" * 60)
debtor_counts = df_raw['Debtor Status'].value_counts(dropna=False)
print(debtor_counts.to_string())

# ── 3. AGE BRACKET — INVESTIGATE ZERO VALUES ─────────────
# Minimum age bracket was 0 — a patient aged 0 days is unusual
# in a billing context. Let us see how many zeros exist
# and what their associated billing looks like
print("\n" + "=" * 60)
print("3. AGE BRACKET — VALUE DISTRIBUTION")
print("=" * 60)
print(df_raw['Age Bracket'].value_counts().sort_index().to_string())

# ── 4. NEGATIVE FINANCIAL VALUES ─────────────────────────
# Total Outstanding, Total Medical Aid Outstanding and
# Total Patient Outstanding all had negative minimums
# Negative outstanding could mean overpayment or data error
# We count how many negative values exist in each column
print("\n" + "=" * 60)
print("4. NEGATIVE VALUES IN FINANCIAL COLUMNS")
print("=" * 60)
financial_cols = [
    'Total Outstanding',
    'Total Medical Aid Outstanding',
    'Total Patient Outstanding',
    'Total Turnover',
    'Total Cashflow'
]
for col in financial_cols:
    neg_count = (df_raw[col] < 0).sum()
    neg_pct = neg_count / len(df_raw) * 100
    print(f"  {col:<35} {neg_count:>5} negative values ({neg_pct:.2f}%)")

# ── 5. OPTION COLUMN — ALL 6 CATEGORIES ──────────────────
# Option had 6 unique values — we saw in R that most records
# are PRIVATE PATIENT ACUTE but some have insurance designations
# Let us confirm the full breakdown here
print("\n" + "=" * 60)
print("5. OPTION COLUMN — ALL CATEGORIES")
print("=" * 60)
print(df_raw['Option'].value_counts(dropna=False).to_string())

1. DUPLICATE INVOICE NUMBERS
Rows sharing an Invoice Number : 22,733
Unique duplicated numbers       : 7,535

Example — top 5 most repeated Invoice Numbers:
Invoice Number
00000001/X    52
00000002/X    22
00000003/X    19
00000004/X    16
00000019/X    15
Name: count, dtype: int64

2. DEBTOR STATUS — ALL UNIQUE VALUES
Debtor Status
Normal                            47120
Payment arrangement                 845
Deceased                            589
To be handed over                   380
Handed over ACK                     364
Handed over MEDICOL                 177
WCA                                 132
Handed over OTHER                   102
private patient                      81
Discontinued                         73
Final point of contact               68
RAF                                  21
Estate                               18
DO NOT USE                           12
ITC listed                            9
Check Notes                           4
No Balance Billing       

In [9]:
# ============================================================
# CELL 5A: EXAMINE DATE VARIABLES
# We look at all 4 date columns to understand their range,
# check for impossible dates, and spot any anomalies.
# Invoice Date is especially important — we know 294 are missing.
# ============================================================

date_cols = ['Invoice Date', 'Capture Date',
             'Earliest Capture Date', 'Latest Capture Date']

for col in date_cols:
    print("=" * 60)
    print(f"VARIABLE: {col}")
    print("=" * 60)

    # Basic stats — min and max tell us the date range
    print(f"  Missing        : {df_raw[col].isnull().sum():,} "
          f"({df_raw[col].isnull().mean()*100:.2f}%)")
    print(f"  Earliest date  : {df_raw[col].min()}")
    print(f"  Latest date    : {df_raw[col].max()}")

    # Year distribution — helps us see if data spans expected range
    print(f"  Records per year:")
    year_counts = df_raw[col].dt.year.value_counts().sort_index()
    for year, count in year_counts.items():
        print(f"    {year}: {count:,}")
    print()

# ── Check logical consistency between date columns ────────
# Capture Date should never be before Invoice Date
# Earliest Capture Date should never be after Latest Capture Date
print("=" * 60)
print("DATE LOGIC CHECKS")
print("=" * 60)

# How many invoices were captured before they were invoiced?
capture_before_invoice = (
    df_raw['Capture Date'] < df_raw['Invoice Date']
).sum()
print(f"  Capture Date before Invoice Date        : {capture_before_invoice:,}")

# How many have Earliest Capture after Latest Capture?
earliest_after_latest = (
    df_raw['Earliest Capture Date'] > df_raw['Latest Capture Date']
).sum()
print(f"  Earliest Capture after Latest Capture   : {earliest_after_latest:,}")

# How many have Latest Capture before Invoice Date?
# This would make our target variable negative — impossible
latest_before_invoice = (
    df_raw['Latest Capture Date'] < df_raw['Invoice Date']
).sum()
print(f"  Latest Capture Date before Invoice Date : {latest_before_invoice:,}")

# How many have Latest Capture before Capture Date?
latest_before_capture = (
    df_raw['Latest Capture Date'] < df_raw['Capture Date']
).sum()
print(f"  Latest Capture Date before Capture Date : {latest_before_capture:,}")

VARIABLE: Invoice Date
  Missing        : 294 (0.59%)
  Earliest date  : 2013-02-18 00:00:00
  Latest date    : 2030-05-20 00:00:00
  Records per year:
    2013.0: 28
    2015.0: 2
    2016.0: 1
    2017.0: 8
    2018.0: 18
    2019.0: 306
    2020.0: 7,511
    2021.0: 9,867
    2022.0: 6,470
    2023.0: 9,085
    2024.0: 16,216
    2025.0: 192
    2030.0: 2

VARIABLE: Capture Date
  Missing        : 0 (0.00%)
  Earliest date  : 2020-01-02 00:00:00
  Latest date    : 2025-01-08 00:00:00
  Records per year:
    2020: 7,646
    2021: 9,664
    2022: 6,550
    2023: 9,259
    2024: 16,603
    2025: 278

VARIABLE: Earliest Capture Date
  Missing        : 0 (0.00%)
  Earliest date  : 2017-07-17 00:00:00
  Latest date    : 2025-01-08 00:00:00
  Records per year:
    2017: 9
    2018: 44
    2019: 77
    2020: 7,730
    2021: 9,999
    2022: 6,389
    2023: 9,148
    2024: 16,326
    2025: 278

VARIABLE: Latest Capture Date
  Missing        : 0 (0.00%)
  Earliest date  : 2020-01-03 00:00:00
 

In [10]:
# ============================================================
# CELL 5B: EXAMINE CATEGORICAL VARIABLES — GROUP 1
# We examine: Option, Gender, Specialty, Debtor Status,
# POSSIBLE PMB CALCULATED, CONTRACTED/NON-CONTRACTED
# For each variable we look at all unique values and their
# counts so we can spot inconsistencies, rare categories,
# capitalisation issues and anything unexpected.
# ============================================================

cat_vars_group1 = [
    'Option',
    'Gender',
    'Specialty',
    'Debtor Status',
    'POSSIBLE PMB CALCULATED',
    'CONTRACTED/NON-CONTRACTED'
]

for col in cat_vars_group1:
    print("=" * 60)
    print(f"VARIABLE: {col}")
    print("=" * 60)
    print(f"  Missing : {df_raw[col].isnull().sum():,} "
          f"({df_raw[col].isnull().mean()*100:.2f}%)")
    print(f"  Unique  : {df_raw[col].nunique()}")
    print(f"\n  Value counts (all):")

    # Show every unique value with count and percentage
    # dropna=False means we also count NaN as a category
    vc = df_raw[col].value_counts(dropna=False)
    for val, count in vc.items():
        pct = count / len(df_raw) * 100
        print(f"    {str(val):<45} {count:>6} ({pct:.2f}%)")
    print()

VARIABLE: Option
  Missing : 0 (0.00%)
  Unique  : 6

  Value counts (all):
    PRIVATE PATIENT ACUTE                          49304 (98.61%)
    PRIVATE PATIENT COST ACUTE                       440 (0.88%)
    PRIVATE PATIENT PRIVATE PATIENT SEP ACUT         190 (0.38%)
    PRIVATE ACUTE                                     47 (0.09%)
    PRIVATE INSURANCE RATE                            15 (0.03%)
    PRIVATE PATIENT PRIVATE PATIENT PHARMACY           4 (0.01%)

VARIABLE: Gender
  Missing : 1,393 (2.79%)
  Unique  : 3

  Value counts (all):
    F                                              25229 (50.46%)
    M                                              23311 (46.62%)
    nan                                             1393 (2.79%)
    V                                                 67 (0.13%)

VARIABLE: Specialty
  Missing : 1 (0.00%)
  Unique  : 118

  Value counts (all):
    (026) Ophthalmologist                           5929 (11.86%)
    (042) General Surgeon                 

In [11]:
# ============================================================
# CELL 5C: EXAMINE CATEGORICAL VARIABLES — GROUP 2
# We examine: Service Centre, Posted Billing Group,
# Primary ICD10, Patient DOB
# Service Centre and Posted Billing Group have high cardinality
# (many unique values) so we look at the top categories and
# the distribution to understand what consolidation is needed.
# Primary ICD10 has 2,387 unique codes — we examine how they
# distribute across chapters.
# Patient DOB loaded as object (not date) — we investigate why.
# ============================================================

# ── Service Centre ────────────────────────────────────────
print("=" * 60)
print("VARIABLE: Service Centre")
print("=" * 60)
print(f"  Missing : {df_raw['Service Centre'].isnull().sum():,} "
      f"({df_raw['Service Centre'].isnull().mean()*100:.2f}%)")
print(f"  Unique  : {df_raw['Service Centre'].nunique()}")
print(f"\n  Top 20 values:")
vc = df_raw['Service Centre'].value_counts(dropna=False)
for val, count in vc.head(20).items():
    pct = count / len(df_raw) * 100
    print(f"    {str(val):<50} {count:>5} ({pct:.2f}%)")
print(f"\n  Categories with only 1 record: "
      f"{(vc == 1).sum()}")
print(f"  Categories with fewer than 10 records: "
      f"{(vc < 10).sum()}")

# ── Posted Billing Group ──────────────────────────────────
print("\n" + "=" * 60)
print("VARIABLE: Posted Billing Group")
print("=" * 60)
print(f"  Missing : {df_raw['Posted Billing Group'].isnull().sum():,} "
      f"({df_raw['Posted Billing Group'].isnull().mean()*100:.2f}%)")
print(f"  Unique  : {df_raw['Posted Billing Group'].nunique()}")
print(f"\n  All values:")
vc_pbg = df_raw['Posted Billing Group'].value_counts(dropna=False)
for val, count in vc_pbg.items():
    pct = count / len(df_raw) * 100
    print(f"    {str(val):<50} {count:>5} ({pct:.2f}%)")

# ── Primary ICD10 ─────────────────────────────────────────
# ICD10 codes follow a letter+number pattern where the first
# letter indicates the disease chapter (A=infectious diseases,
# C=cancer, J=respiratory, etc.)
# We extract just the first character to see chapter distribution
print("\n" + "=" * 60)
print("VARIABLE: Primary ICD10")
print("=" * 60)
print(f"  Missing : {df_raw['Primary ICD10'].isnull().sum():,} "
      f"({df_raw['Primary ICD10'].isnull().mean()*100:.2f}%)")
print(f"  Unique  : {df_raw['Primary ICD10'].nunique()}")
print(f"\n  Distribution by ICD10 chapter (first letter):")
icd_chapter = df_raw['Primary ICD10'].str[0].value_counts(dropna=False)
for chapter, count in icd_chapter.items():
    pct = count / len(df_raw) * 100
    print(f"    Chapter {str(chapter):<5} {count:>5} ({pct:.2f}%)")

# ── Patient DOB ───────────────────────────────────────────
# This loaded as object instead of datetime — we investigate
# what values are causing it to fail date parsing
print("\n" + "=" * 60)
print("VARIABLE: Patient DOB")
print("=" * 60)
print(f"  Missing : {df_raw['Patient DOB'].isnull().sum():,} "
      f"({df_raw['Patient DOB'].isnull().mean()*100:.2f}%)")
print(f"  Unique  : {df_raw['Patient DOB'].nunique()}")

# Try converting to date and see how many fail
dob_converted = pd.to_datetime(df_raw['Patient DOB'], errors='coerce')
failed = dob_converted.isnull().sum() - df_raw['Patient DOB'].isnull().sum()
print(f"  Values that fail date conversion: {failed:,}")
print(f"  Earliest DOB (if converted)     : {dob_converted.min()}")
print(f"  Latest DOB (if converted)       : {dob_converted.max()}")

# Calculate implied ages to spot impossible values
# Anyone born after 2025 or before 1900 is suspicious
current_year = 2026
dob_years = dob_converted.dt.year
print(f"\n  DOB year distribution (suspicious ranges):")
print(f"    Born after 2025 (future)  : "
      f"{(dob_years > 2025).sum():,}")
print(f"    Born before 1900          : "
      f"{(dob_years < 1900).sum():,}")
print(f"    Plausible range 1900-2025 : "
      f"{((dob_years >= 1900) & (dob_years <= 2025)).sum():,}")

VARIABLE: Service Centre
  Missing : 2,115 (4.23%)
  Unique  : 730

  Top 20 values:
    ROOMS - KAMERS                                      3599 (7.20%)
    nan                                                 2115 (4.23%)
    KAMERS/ROOMS                                        1824 (3.65%)
    HOSPITAL - HOSPITAAL                                1334 (2.67%)
    DENTISTS AT MEMORIAL ROOMS                           972 (1.94%)
    CONSTANTIABERG ROOMS                                 873 (1.75%)
    ROOMS                                                808 (1.62%)
    DENTISTS AT MEMORIAL                                 711 (1.42%)
    NELSPRUIT EYE INSTITUTE                              668 (1.34%)
    THE OXFORD CENTRE ROOMS                              648 (1.30%)
    CHRISTIAAN BARNARD MEMORIAL HOSPITAL                 572 (1.14%)
    BLAAUWBERG ROOMS                                     571 (1.14%)
    LIFE CARSTENHOF HOSPITAL                             554 (1.11%)
    ST AUGUSTINES 

In [12]:
# ============================================================
# CELL 5D: EXAMINE NUMERIC VARIABLES
# We examine: Age Bracket, Total Outstanding,
# Total Medical Aid Outstanding, Total Patient Outstanding,
# Total Turnover, Total Cashflow, Total Journals
# Age Bracket is numeric but actually behaves like a category
# The financial variables we examine for negative values,
# extreme outliers, and relationships between each other.
# ============================================================

# ── Age Bracket ───────────────────────────────────────────
# We saw 7 distinct values (0,30,60,90,120,150,180)
# This is a debt aging bucket — how long the debt has been
# outstanding in 30-day intervals. Behaves like ordered category.
print("=" * 60)
print("VARIABLE: Age Bracket")
print("=" * 60)
print(f"  Missing : {df_raw['Age Bracket'].isnull().sum():,}")
print(f"  Unique  : {df_raw['Age Bracket'].nunique()}")
print(f"\n  All values with counts:")
ab_counts = df_raw['Age Bracket'].value_counts().sort_index()
for val, count in ab_counts.items():
    pct = count / len(df_raw) * 100
    bar = "█" * int(pct / 1)
    print(f"    {val:>4} days : {count:>6} ({pct:.2f}%) {bar}")

# ── Financial variables ───────────────────────────────────
# We look at each financial column in detail
# Key questions: how many zeros, how many negatives,
# how extreme are the outliers, and are columns related?
financial_vars = [
    'Total Outstanding',
    'Total Medical Aid Outstanding',
    'Total Patient Outstanding',
    'Total Turnover',
    'Total Cashflow',
    'Total Journals'
]

print("\n" + "=" * 60)
print("FINANCIAL VARIABLES — DETAILED EXAMINATION")
print("=" * 60)

for col in financial_vars:
    print(f"\n  --- {col} ---")
    print(f"  Missing  : {df_raw[col].isnull().sum():,} "
          f"({df_raw[col].isnull().mean()*100:.2f}%)")

    non_null = df_raw[col].dropna()
    print(f"  Zero     : {(non_null == 0).sum():,} "
          f"({(non_null == 0).mean()*100:.2f}% of non-missing)")
    print(f"  Negative : {(non_null < 0).sum():,} "
          f"({(non_null < 0).mean()*100:.2f}% of non-missing)")
    print(f"  Positive : {(non_null > 0).sum():,} "
          f"({(non_null > 0).mean()*100:.2f}% of non-missing)")
    print(f"  Min      : {non_null.min():>15,.2f}")
    print(f"  Max      : {non_null.max():>15,.2f}")
    print(f"  Mean     : {non_null.mean():>15,.2f}")
    print(f"  Median   : {non_null.median():>15,.2f}")

# ── Relationship check between financial columns ──────────
# Total Outstanding should equal Total Turnover + Total Cashflow
# + Total Journals (since cashflow is negative and journals
# are adjustments). Let us verify this relationship holds.
print("\n" + "=" * 60)
print("FINANCIAL RELATIONSHIP CHECK")
print("=" * 60)

# Only check rows where all four columns have values
complete_financial = df_raw[[
    'Total Outstanding', 'Total Turnover',
    'Total Cashflow', 'Total Journals'
]].dropna()

# Calculate expected outstanding
complete_financial = complete_financial.copy()
complete_financial['Expected Outstanding'] = (
    complete_financial['Total Turnover'] +
    complete_financial['Total Cashflow'] +
    complete_financial['Total Journals']
)
complete_financial['Difference'] = (
    complete_financial['Total Outstanding'] -
    complete_financial['Expected Outstanding']
).abs()

# How many rows match the formula exactly (within rounding)?
matches = (complete_financial['Difference'] < 0.01).sum()
total   = len(complete_financial)
print(f"  Rows where Outstanding = Turnover + Cashflow + Journals:")
print(f"  {matches:,} out of {total:,} ({matches/total*100:.2f}%)")
print(f"\n  Max discrepancy : {complete_financial['Difference'].max():,.2f}")
print(f"  Mean discrepancy: {complete_financial['Difference'].mean():,.2f}")

VARIABLE: Age Bracket
  Missing : 0
  Unique  : 7

  All values with counts:
       0 days :    278 (0.56%) 
      30 days :   1449 (2.90%) ██
      60 days :   1729 (3.46%) ███
      90 days :   2047 (4.09%) ████
     120 days :   1621 (3.24%) ███
     150 days :   1923 (3.85%) ███
     180 days :  40953 (81.91%) █████████████████████████████████████████████████████████████████████████████████

FINANCIAL VARIABLES — DETAILED EXAMINATION

  --- Total Outstanding ---
  Missing  : 0 (0.00%)
  Zero     : 48,505 (97.01% of non-missing)
  Negative : 186 (0.37% of non-missing)
  Positive : 1,309 (2.62% of non-missing)
  Min      :     -110,735.80
  Max      :       75,700.00
  Mean     :           45.27
  Median   :            0.00

  --- Total Medical Aid Outstanding ---
  Missing  : 0 (0.00%)
  Zero     : 49,874 (99.75% of non-missing)
  Negative : 73 (0.15% of non-missing)
  Positive : 53 (0.11% of non-missing)
  Min      :      -85,628.90
  Max      :       19,106.29
  Mean     :        

In [13]:
# ============================================================
# CELL 5E: INVESTIGATE ANOMALOUS CASHFLOW RECORDS
# Total Cashflow should always be negative (money coming in
# shown as credit) or zero. 37 records have positive cashflow
# which is unexpected. We examine these to understand why.
# ============================================================

# Filter to positive cashflow records only
positive_cashflow = df_raw[df_raw['Total Cashflow'] > 0][[
    'Invoice Number', 'Invoice Date', 'Capture Date',
    'Total Turnover', 'Total Cashflow',
    'Total Outstanding', 'Total Journals',
    'Debtor Status', 'Option'
]]

print("=" * 60)
print("POSITIVE CASHFLOW RECORDS (37 records)")
print("=" * 60)
print(f"  Count : {len(positive_cashflow)}")
print(f"\n  Debtor Status breakdown:")
print(positive_cashflow['Debtor Status'].value_counts().to_string())
print(f"\n  Option breakdown:")
print(positive_cashflow['Option'].value_counts().to_string())
print(f"\n  Financial summary:")
print(positive_cashflow[['Total Turnover','Total Cashflow',
                          'Total Outstanding','Total Journals']]\
      .describe().round(2).to_string())
print(f"\n  All records:")
pd.set_option('display.max_rows', 40)
print(positive_cashflow.to_string(index=False))

POSITIVE CASHFLOW RECORDS (37 records)
  Count : 37

  Debtor Status breakdown:
Debtor Status
Normal    37

  Option breakdown:
Option
PRIVATE PATIENT ACUTE    37

  Financial summary:
       Total Turnover  Total Cashflow  Total Outstanding  Total Journals
count            8.00           37.00              37.00           16.00
mean           168.75         5348.26            3259.99        -4913.50
std            315.03         7981.49            4962.57        10607.33
min              0.00            3.20               0.00       -43309.18
25%              0.00          720.14               0.00        -2850.75
50%              0.00         1708.83             397.50        -1619.40
75%            150.00         8979.20            7049.80         -930.58
max            750.00        43309.18           19106.29           -3.20

  All records:
Invoice Number Invoice Date Capture Date  Total Turnover  Total Cashflow  Total Outstanding  Total Journals Debtor Status                Optio

In [14]:
# ============================================================
# CELL 6: COMPLETE VARIABLE DESCRIPTION TABLE
# Section 3.1 - Data Understanding
# Now that we have examined every variable individually we
# can document them accurately. This table will feed directly
# into the report. It summarises every variable's type,
# missingness, unique value count and our own description
# based on what we actually observed — not assumed.
# ============================================================

# Each entry: (Type, Role, Description based on our findings)
# Role = whether we expect to KEEP or EXCLUDE in preparation
# This is just an early flag — final decision made in 3.2.1

variable_info = {
    'Invoice Number':               ('Categorical', 'Investigate',
        'Invoice identifier — not unique, 7,535 numbers appear in multiple rows across 22,733 records'),
    'Entity name':                  ('Categorical', 'Exclude',
        'Hashed practice identifier — anonymised, no predictive value'),
    'Main Member':                  ('Categorical', 'Exclude',
        'Hashed member identifier — anonymised, no predictive value'),
    'Medical Aid Number':           ('Categorical', 'Exclude',
        'Hashed medical aid number — anonymised, no predictive value'),
    'Invoice Date':                 ('DateTime',    'Investigate',
        'Service invoice date — 294 missing, 2 records dated 2030 (impossible), sparse pre-2019'),
    'Capture Date':                 ('DateTime',    'Keep',
        'Date invoice entered billing system — no missing, used as target variable start point'),
    'Earliest Capture Date':        ('DateTime',    'Exclude',
        'Earliest capture across billing events — used only to understand processing history'),
    'Latest Capture Date':          ('DateTime',    'Keep',
        'Latest capture date — used as target variable end point'),
    'Service Centre':               ('Categorical', 'Engineer',
        '730 unique values — 111 appear once, to be grouped into facility type categories'),
    'Treating Doctor':              ('Categorical', 'Exclude',
        'Hashed doctor identifier — anonymised, 566 unique values, no predictive value'),
    'Referring Doctor Name':        ('Categorical', 'Engineer',
        'Hashed referral identifier — to be converted to binary Has Referral (0/1)'),
    'Scheme':                       ('Categorical', 'Exclude',
        'Constant — single value PRIVATE PATIENT across all 50,000 records'),
    'Option':                       ('Categorical', 'Keep',
        '6 categories — 98.61% PRIVATE PATIENT ACUTE, minor categories to be reviewed'),
    'Administrator':                ('Categorical', 'Exclude',
        'Constant — single value PRIVATE across all 50,000 records'),
    'Posted Billing Group':         ('Categorical', 'Keep',
        '75 values — spacing inconsistencies confirmed, unknown and RESERVED labels need cleaning'),
    'ICD10 String':                 ('Categorical', 'Exclude',
        'Full ICD10 text string — duplicate of Primary ICD10, noisier format'),
    'Primary ICD10':                ('Categorical', 'Engineer',
        '2,387 unique codes — U/u capitalisation issue found, to be grouped into ICD10 chapters'),
    'ICD10 Description':            ('Categorical', 'Exclude',
        'Text description of ICD10 — duplicate information, harder to use than code'),
    'Patient Name':                 ('Categorical', 'Exclude',
        'Hashed patient identifier — anonymised, 468 missing, no predictive value'),
    'Patient DOB':                  ('Categorical', 'Exclude',
        '42.48% missing, earliest date is year 0101 — corrupted placeholder dates, unusable'),
    'Gender':                       ('Categorical', 'Keep',
        '3 values: F (50.46%), M (46.62%), V (0.13%) — V to be consolidated into F, 2.79% missing'),
    'Specialty':                    ('Categorical', 'Keep',
        '118 values — confirmed duplicates, typos, bare numbers — needs standardisation'),
    'Debtor Status':                ('Categorical', 'Keep',
        '20 values — confirmed inconsistencies: private patient, ESTATE/Estate, WCA, RAF, DO NOT USE'),
    'POSSIBLE PMB CALCULATED':      ('Categorical', 'Keep',
        'Clean — 2 values: NON-PMB (50.58%) and PMB (49.42%), no missing'),
    'CONTRACTED/NON-CONTRACTED':    ('Categorical', 'Keep',
        'Clean — 2 values: CONTRACTED (54.10%) and NON-CONTRACTED (45.90%), no missing'),
    'Age Bracket':                  ('Numeric',     'Keep',
        '7 ordered values (0-180 days) — 81.91% at 180 days, treat as ordered categorical'),
    'Total Outstanding':            ('Numeric',     'Exclude',
        '97.01% zero, 186 negative values — post-capture outcome variable, not a predictor'),
    'Total Medical Aid Outstanding':('Numeric',     'Exclude',
        '99.75% zero — near constant, post-capture outcome variable'),
    'Total Patient Outstanding':    ('Numeric',     'Exclude',
        '97.24% zero, 118 negative values — post-capture outcome variable'),
    'Total Turnover':               ('Numeric',     'Keep',
        '7.64% missing, heavily right-skewed (max R1.48M) — needs log transformation'),
    'Total Cashflow':               ('Numeric',     'Exclude',
        '90.43% negative by accounting convention — post-capture outcome, 37 anomalous positives noted'),
    'Total Journals':               ('Numeric',     'Exclude',
        '93.67% missing — post-capture write-off adjustments, not available at time of capture'),
}

# ── Build the summary DataFrame ───────────────────────────
rows = []
for col, (var_type, role, desc) in variable_info.items():
    missing_n   = df_raw[col].isnull().sum()
    missing_pct = round(missing_n / len(df_raw) * 100, 2)
    unique_n    = df_raw[col].nunique()
    rows.append({
        'Variable':      col,
        'Type':          var_type,
        'Unique':        unique_n,
        'Missing N':     missing_n,
        'Missing %':     missing_pct,
        'Initial Role':  role,
        'Description':   desc
    })

variable_table = pd.DataFrame(rows)

# ── Print formal table ────────────────────────────────────
print("=" * 60)
print("COMPLETE VARIABLE DESCRIPTION TABLE")
print(f"Source: PRIVATE PATIENT.xlsx")
print(f"Dimensions: {df_raw.shape[0]:,} rows x {df_raw.shape[1]} columns")
print("=" * 60)
pd.set_option('display.max_colwidth', 85)
pd.set_option('display.max_rows', 40)
print(variable_table.to_string(index=False))

# ── Summary counts ────────────────────────────────────────
print("\n" + "=" * 60)
print("SECTION 3.1 SUMMARY COUNTS")
print("=" * 60)
print(f"\n  By variable type:")
for t, n in variable_table['Type'].value_counts().items():
    print(f"    {t:<15} : {n}")
print(f"\n  By initial role:")
for r, n in variable_table['Initial Role'].value_counts().items():
    print(f"    {r:<15} : {n}")
print(f"\n  Total variables  : {len(variable_table)}")
print(f"  Total records    : {df_raw.shape[0]:,}")
print(f"\n  Variables with missing values : "
      f"{(variable_table['Missing N'] > 0).sum()}")
print(f"  Variables with no missing     : "
      f"{(variable_table['Missing N'] == 0).sum()}")

# ── Save the variable table to outputs ────────────────────
# We save this as CSV so we can reference it in the report
# and in the data preparation notebook
variable_table.to_csv(
    r'C:\2026\MedicalBillingResearch\outputs\tables\variable_description_table.csv',
    index=False
)
print(f"\n  Variable description table saved to outputs/tables/")

COMPLETE VARIABLE DESCRIPTION TABLE
Source: PRIVATE PATIENT.xlsx
Dimensions: 50,000 rows x 32 columns
                     Variable        Type  Unique  Missing N  Missing % Initial Role                                                                                   Description
               Invoice Number Categorical   34802          0       0.00  Investigate  Invoice identifier — not unique, 7,535 numbers appear in multiple rows across 22,733 records
                  Entity name Categorical     397          0       0.00      Exclude                                  Hashed practice identifier — anonymised, no predictive value
                  Main Member Categorical   22084          0       0.00      Exclude                                    Hashed member identifier — anonymised, no predictive value
           Medical Aid Number Categorical    1584        145       0.29      Exclude                                   Hashed medical aid number — anonymised, no predictive value
   

## System Migration Investigation
### Why Records Pre-2020 Matter

Invoice Date starts from 2013 but Capture Date starts from 2020.
This gap suggests a billing system migration occurred around 2020.
Pre-2020 invoice records may be historical imports rather than
live billing events processed through the current system.

We investigate this before committing to any target variable
definition or threshold in data preparation.

In [15]:
# ============================================================
# CELL 7: SYSTEM MIGRATION INVESTIGATION
# Invoice Date goes back to 2013 but Capture Date only starts
# from 2020. This suggests a system migration happened around
# 2020 where historical records were imported into the new
# billing system. We investigate whether pre-2020 Invoice Date
# records behave differently from post-2020 records because
# if they are just historical imports they may not represent
# genuine processing events in the current system.
# ============================================================

# Reload raw data with Invoice Date since we need it here
df_inv = pd.read_excel(
    r'C:\2026\MedicalBillingResearch\dataset\PRIVATE PATIENT.xlsx',
    engine='openpyxl',
    usecols=['Invoice Number', 'Invoice Date',
             'Capture Date', 'Latest Capture Date']
)

# ── Basic split — pre vs post 2020 Invoice Date ───────────
pre_2020  = df_inv[df_inv['Invoice Date'].dt.year < 2020]
post_2020 = df_inv[df_inv['Invoice Date'].dt.year >= 2020]
missing_inv = df_inv['Invoice Date'].isnull()

print("=" * 60)
print("INVOICE DATE — PRE vs POST 2020 SPLIT")
print("=" * 60)
print(f"  Invoice Date before 2020  : {len(pre_2020):,} records "
      f"({len(pre_2020)/len(df_inv)*100:.2f}%)")
print(f"  Invoice Date 2020 onwards : {len(post_2020):,} records "
      f"({len(post_2020)/len(df_inv)*100:.2f}%)")
print(f"  Invoice Date missing      : {missing_inv.sum():,} records "
      f"({missing_inv.sum()/len(df_inv)*100:.2f}%)")

# ── Calculate processing time for both groups ─────────────
# Using our formula: Latest Capture Date - Capture Date
# This tells us how long each invoice spent inside the system
df_inv['Processing_Days'] = (
    df_inv['Latest Capture Date'] - df_inv['Capture Date']
).dt.days

# Also calculate the full journey for comparison
# Latest Capture Date - Invoice Date
df_inv['Full_Journey_Days'] = (
    df_inv['Latest Capture Date'] - df_inv['Invoice Date']
).dt.days

print("\n" + "=" * 60)
print("PROCESSING TIME — Latest Capture Date minus Capture Date")
print("=" * 60)

for label, subset in [
    ('Pre-2020 Invoice Date',  df_inv[df_inv['Invoice Date'].dt.year < 2020]),
    ('Post-2020 Invoice Date', df_inv[df_inv['Invoice Date'].dt.year >= 2020]),
    ('All records',            df_inv)
]:
    days = subset['Processing_Days'].dropna()
    print(f"\n  {label} ({len(subset):,} records):")
    print(f"    Min    : {days.min():.0f} days")
    print(f"    Median : {days.median():.0f} days")
    print(f"    Mean   : {days.mean():.1f} days")
    print(f"    Max    : {days.max():.0f} days")
    print(f"    Zero   : {(days == 0).sum():,} records "
          f"({(days == 0).mean()*100:.1f}%)")
    print(f"    ≤14    : {(days <= 14).sum():,} records "
          f"({(days <= 14).mean()*100:.1f}%)")
    print(f"    ≤30    : {(days <= 30).sum():,} records "
          f"({(days <= 30).mean()*100:.1f}%)")
    print(f"    >30    : {(days > 30).sum():,} records "
          f"({(days > 30).mean()*100:.1f}%)")

# ── Full journey comparison for pre-2020 records ──────────
# If pre-2020 records are historical imports, their full
# journey days will be enormous (7+ years) but their
# processing days inside the system will be very small
print("\n" + "=" * 60)
print("FULL JOURNEY — Latest Capture Date minus Invoice Date")
print("PRE-2020 RECORDS ONLY")
print("=" * 60)
pre = df_inv[df_inv['Invoice Date'].dt.year < 2020].copy()
print(f"  Min full journey    : {pre['Full_Journey_Days'].min():.0f} days")
print(f"  Median full journey : {pre['Full_Journey_Days'].median():.0f} days")
print(f"  Mean full journey   : {pre['Full_Journey_Days'].mean():.1f} days")
print(f"  Max full journey    : {pre['Full_Journey_Days'].max():.0f} days")
print(f"\n  Min in-system days  : {pre['Processing_Days'].min():.0f} days")
print(f"  Median in-system    : {pre['Processing_Days'].median():.0f} days")
print(f"  Mean in-system      : {pre['Processing_Days'].mean():.1f} days")
print(f"  Max in-system       : {pre['Processing_Days'].max():.0f} days")

# ── Threshold sensitivity — what split does each give us ──
# We test multiple thresholds on ALL records to see which
# produces a meaningful and defensible class split
print("\n" + "=" * 60)
print("THRESHOLD SENSITIVITY — ALL 50,000 RECORDS")
print("What class split does each threshold produce?")
print("=" * 60)
all_days = df_inv['Processing_Days'].dropna()
thresholds = [7, 10, 14, 21, 30, 45, 60]
print(f"  {'Threshold':<12} {'Timely':>10} {'Delayed':>10} "
      f"{'Timely %':>10} {'Delayed %':>10}")
print(f"  {'-'*52}")
for t in thresholds:
    timely  = (all_days <= t).sum()
    delayed = (all_days > t).sum()
    print(f"  {t:<12} {timely:>10,} {delayed:>10,} "
          f"{timely/len(all_days)*100:>9.1f}% "
          f"{delayed/len(all_days)*100:>9.1f}%")

INVOICE DATE — PRE vs POST 2020 SPLIT
  Invoice Date before 2020  : 363 records (0.73%)
  Invoice Date 2020 onwards : 49,343 records (98.69%)
  Invoice Date missing      : 294 records (0.59%)

PROCESSING TIME — Latest Capture Date minus Capture Date

  Pre-2020 Invoice Date (363 records):
    Min    : 0 days
    Median : 7 days
    Mean   : 81.5 days
    Max    : 1179 days
    Zero   : 159 records (43.8%)
    ≤14    : 206 records (56.7%)
    ≤30    : 225 records (62.0%)
    >30    : 138 records (38.0%)

  Post-2020 Invoice Date (49,343 records):
    Min    : 0 days
    Median : 0 days
    Mean   : 26.2 days
    Max    : 1640 days
    Zero   : 27,655 records (56.0%)
    ≤14    : 37,489 records (76.0%)
    ≤30    : 41,379 records (83.9%)
    >30    : 7,964 records (16.1%)

  All records (50,000 records):
    Min    : 0 days
    Median : 0 days
    Mean   : 26.5 days
    Max    : 1640 days
    Zero   : 28,041 records (56.1%)
    ≤14    : 37,962 records (75.9%)
    ≤30    : 41,883 records 

## Zero Processing Time Investigation

56.1% of all records show zero processing time under the formula
Latest Capture Date minus Capture Date. This means Latest Capture
Date equals Capture Date exactly for 28,041 records.

We investigate whether these same-day records represent genuine
invoice resolution or incomplete system recording by examining
their Debtor Status and financial profiles.

In [2]:
# ============================================================
# CELL 8: ZERO PROCESSING TIME INVESTIGATION
# Section 3.1 - Data Understanding
# 56.1% of records have Latest Capture Date = Capture Date
# meaning zero days of processing time inside the system.
# This is a very high proportion. We investigate whether
# these represent genuine same-day resolution or incomplete
# system recording by comparing zero-day vs non-zero-day
# records across key variables.
# If zero-day records are overwhelmingly Normal status with
# zero outstanding balance that supports genuine resolution.
# If some have Active Collection or Payment Arrangement
# status that suggests the system stopped recording but
# the invoice was not actually resolved.
# ============================================================

# Reload raw data with the columns we need for this check
# We need date columns and financial/status columns together
df_inv2 = pd.read_excel(
    r'C:\2026\MedicalBillingResearch\dataset\PRIVATE PATIENT.xlsx',
    engine='openpyxl',
    usecols=[
        'Invoice Number', 'Capture Date', 'Latest Capture Date',
        'Debtor Status', 'Total Outstanding', 'Total Turnover',
        'Total Cashflow', 'Specialty', 'Service Centre'
    ]
)

# Calculate processing time
df_inv2['Processing Days'] = (
    df_inv2['Latest Capture Date'] - df_inv2['Capture Date']
).dt.days

# Split into zero-day and non-zero-day groups
zero_day     = df_inv2[df_inv2['Processing Days'] == 0]
non_zero_day = df_inv2[df_inv2['Processing Days'] > 0]

print("=" * 60)
print("ZERO vs NON-ZERO PROCESSING TIME — OVERVIEW")
print("=" * 60)
print(f"\n  Zero-day records     : {len(zero_day):,} "
      f"({len(zero_day)/len(df_inv2)*100:.1f}%)")
print(f"  Non-zero-day records : {len(non_zero_day):,} "
      f"({len(non_zero_day)/len(df_inv2)*100:.1f}%)")
print(f"  Total records        : {len(df_inv2):,}")

# ── Debtor Status comparison ──────────────────────────────
# Key question: do zero-day records have different payment
# status profiles compared to non-zero-day records?
# If zero-day records are genuinely resolved we expect them
# to be overwhelmingly Normal status
print(f"\n{'=' * 60}")
print("DEBTOR STATUS — ZERO-DAY vs NON-ZERO-DAY")
print("=" * 60)

# Get value counts for both groups
zero_status     = zero_day['Debtor Status'].value_counts()
non_zero_status = non_zero_day['Debtor Status'].value_counts()

# Get all unique statuses across both groups
all_statuses = set(
    zero_status.index.tolist() +
    non_zero_status.index.tolist()
)

print(f"\n  {'Debtor Status':<35} {'Zero-day':>10} "
      f"{'Zero%':>8} {'NonZero':>10} {'NonZero%':>10}")
print(f"  {'-'*75}")

for status in sorted(all_statuses,
                      key=lambda x: zero_status.get(x, 0),
                      reverse=True):
    z_count  = zero_status.get(status, 0)
    nz_count = non_zero_status.get(status, 0)
    z_pct    = z_count / len(zero_day) * 100
    nz_pct   = nz_count / len(non_zero_day) * 100
    print(f"  {str(status):<35} {z_count:>10,} "
          f"{z_pct:>7.1f}% {nz_count:>10,} {nz_pct:>9.1f}%")

# ── Financial profile comparison ──────────────────────────
# If zero-day records are genuinely resolved we expect
# Total Outstanding to be zero — fully paid
# If they are not resolved we expect positive outstanding
print(f"\n{'=' * 60}")
print("FINANCIAL PROFILE — ZERO-DAY vs NON-ZERO-DAY")
print("=" * 60)

for label, group in [
    ('Zero-day records',     zero_day),
    ('Non-zero-day records', non_zero_day)
]:
    outstanding = group['Total Outstanding'].dropna()
    turnover    = group['Total Turnover'].dropna()
    print(f"\n  {label} ({len(group):,} records):")
    print(f"    Outstanding = 0    : "
          f"{(outstanding == 0).sum():,} "
          f"({(outstanding == 0).mean()*100:.1f}%)")
    print(f"    Outstanding > 0    : "
          f"{(outstanding > 0).sum():,} "
          f"({(outstanding > 0).mean()*100:.1f}%)")
    print(f"    Outstanding < 0    : "
          f"{(outstanding < 0).sum():,} "
          f"({(outstanding < 0).mean()*100:.1f}%)")
    print(f"    Turnover missing   : "
          f"{turnover.isnull().sum():,} "
          f"({turnover.isnull().mean()*100:.1f}%)")
    print(f"    Turnover = 0       : "
          f"{(turnover == 0).sum():,} "
          f"({(turnover == 0).mean()*100:.1f}%)")
    print(f"    Mean Turnover      : "
          f"R{turnover.mean():,.2f}")

# ── Cashflow check ────────────────────────────────────────
# If payment was received on the same day as capture
# we expect Total Cashflow to be negative (money came in)
# If cashflow is zero that suggests no payment was recorded
print(f"\n{'=' * 60}")
print("CASHFLOW PROFILE — ZERO-DAY RECORDS ONLY")
print("Were payments actually received on same-day records?")
print("=" * 60)

zero_cashflow = zero_day['Total Cashflow'].dropna()
print(f"\n  Cashflow missing   : "
      f"{zero_day['Total Cashflow'].isnull().sum():,} "
      f"({zero_day['Total Cashflow'].isnull().mean()*100:.1f}%)")
print(f"  Cashflow = 0       : "
      f"{(zero_cashflow == 0).sum():,} "
      f"({(zero_cashflow == 0).mean()*100:.1f}%)")
print(f"  Cashflow < 0       : "
      f"{(zero_cashflow < 0).sum():,} "
      f"({(zero_cashflow < 0).mean()*100:.1f}%)")
print(f"  Cashflow > 0       : "
      f"{(zero_cashflow > 0).sum():,} "
      f"({(zero_cashflow > 0).mean()*100:.1f}%)")
print(f"  Mean Cashflow      : "
      f"R{zero_cashflow.mean():,.2f}")

# ── Summary conclusion ────────────────────────────────────
print(f"\n{'=' * 60}")
print("INTERPRETATION SUMMARY")
print("=" * 60)
print(f"""
  If zero-day records show:
  - Overwhelmingly Normal debtor status
  - Total Outstanding = 0 for most records
  - Negative cashflow (payment received)
  → Supports Explanation 1: Genuine same-day resolution

  If zero-day records show:
  - Mix of problematic debtor statuses
  - Positive outstanding balances
  - Zero or missing cashflow
  → Supports Explanation 2: Incomplete system recording

  This finding must be documented as a data limitation
  in the report regardless of which explanation dominates.
""")

NameError: name 'pd' is not defined

In [3]:
# ============================================================
# RE-IMPORT LIBRARIES
# Pandas needs to be imported again because the kernel was
# restarted or this cell is being run independently.
# This ensures the notebook can run from this point without
# needing to re-run all previous cells.
# ============================================================

import pandas as pd
import numpy as np

In [4]:
# ============================================================
# CELL 8: ZERO PROCESSING TIME INVESTIGATION
# Section 3.1 - Data Understanding
# 56.1% of records have Latest Capture Date = Capture Date
# meaning zero days of processing time inside the system.
# This is a very high proportion. We investigate whether
# these represent genuine same-day resolution or incomplete
# system recording by comparing zero-day vs non-zero-day
# records across key variables.
# If zero-day records are overwhelmingly Normal status with
# zero outstanding balance that supports genuine resolution.
# If some have Active Collection or Payment Arrangement
# status that suggests the system stopped recording but
# the invoice was not actually resolved.
# ============================================================

# Reload raw data with the columns we need for this check
# We need date columns and financial/status columns together
df_inv2 = pd.read_excel(
    r'C:\2026\MedicalBillingResearch\dataset\PRIVATE PATIENT.xlsx',
    engine='openpyxl',
    usecols=[
        'Invoice Number', 'Capture Date', 'Latest Capture Date',
        'Debtor Status', 'Total Outstanding', 'Total Turnover',
        'Total Cashflow', 'Specialty', 'Service Centre'
    ]
)

# Calculate processing time
df_inv2['Processing Days'] = (
    df_inv2['Latest Capture Date'] - df_inv2['Capture Date']
).dt.days

# Split into zero-day and non-zero-day groups
zero_day     = df_inv2[df_inv2['Processing Days'] == 0]
non_zero_day = df_inv2[df_inv2['Processing Days'] > 0]

print("=" * 60)
print("ZERO vs NON-ZERO PROCESSING TIME — OVERVIEW")
print("=" * 60)
print(f"\n  Zero-day records     : {len(zero_day):,} "
      f"({len(zero_day)/len(df_inv2)*100:.1f}%)")
print(f"  Non-zero-day records : {len(non_zero_day):,} "
      f"({len(non_zero_day)/len(df_inv2)*100:.1f}%)")
print(f"  Total records        : {len(df_inv2):,}")

# ── Debtor Status comparison ──────────────────────────────
# Key question: do zero-day records have different payment
# status profiles compared to non-zero-day records?
# If zero-day records are genuinely resolved we expect them
# to be overwhelmingly Normal status
print(f"\n{'=' * 60}")
print("DEBTOR STATUS — ZERO-DAY vs NON-ZERO-DAY")
print("=" * 60)

# Get value counts for both groups
zero_status     = zero_day['Debtor Status'].value_counts()
non_zero_status = non_zero_day['Debtor Status'].value_counts()

# Get all unique statuses across both groups
all_statuses = set(
    zero_status.index.tolist() +
    non_zero_status.index.tolist()
)

print(f"\n  {'Debtor Status':<35} {'Zero-day':>10} "
      f"{'Zero%':>8} {'NonZero':>10} {'NonZero%':>10}")
print(f"  {'-'*75}")

for status in sorted(all_statuses,
                      key=lambda x: zero_status.get(x, 0),
                      reverse=True):
    z_count  = zero_status.get(status, 0)
    nz_count = non_zero_status.get(status, 0)
    z_pct    = z_count / len(zero_day) * 100
    nz_pct   = nz_count / len(non_zero_day) * 100
    print(f"  {str(status):<35} {z_count:>10,} "
          f"{z_pct:>7.1f}% {nz_count:>10,} {nz_pct:>9.1f}%")

# ── Financial profile comparison ──────────────────────────
# If zero-day records are genuinely resolved we expect
# Total Outstanding to be zero — fully paid
# If they are not resolved we expect positive outstanding
print(f"\n{'=' * 60}")
print("FINANCIAL PROFILE — ZERO-DAY vs NON-ZERO-DAY")
print("=" * 60)

for label, group in [
    ('Zero-day records',     zero_day),
    ('Non-zero-day records', non_zero_day)
]:
    outstanding = group['Total Outstanding'].dropna()
    turnover    = group['Total Turnover'].dropna()
    print(f"\n  {label} ({len(group):,} records):")
    print(f"    Outstanding = 0    : "
          f"{(outstanding == 0).sum():,} "
          f"({(outstanding == 0).mean()*100:.1f}%)")
    print(f"    Outstanding > 0    : "
          f"{(outstanding > 0).sum():,} "
          f"({(outstanding > 0).mean()*100:.1f}%)")
    print(f"    Outstanding < 0    : "
          f"{(outstanding < 0).sum():,} "
          f"({(outstanding < 0).mean()*100:.1f}%)")
    print(f"    Turnover missing   : "
          f"{turnover.isnull().sum():,} "
          f"({turnover.isnull().mean()*100:.1f}%)")
    print(f"    Turnover = 0       : "
          f"{(turnover == 0).sum():,} "
          f"({(turnover == 0).mean()*100:.1f}%)")
    print(f"    Mean Turnover      : "
          f"R{turnover.mean():,.2f}")

# ── Cashflow check ────────────────────────────────────────
# If payment was received on the same day as capture
# we expect Total Cashflow to be negative (money came in)
# If cashflow is zero that suggests no payment was recorded
print(f"\n{'=' * 60}")
print("CASHFLOW PROFILE — ZERO-DAY RECORDS ONLY")
print("Were payments actually received on same-day records?")
print("=" * 60)

zero_cashflow = zero_day['Total Cashflow'].dropna()
print(f"\n  Cashflow missing   : "
      f"{zero_day['Total Cashflow'].isnull().sum():,} "
      f"({zero_day['Total Cashflow'].isnull().mean()*100:.1f}%)")
print(f"  Cashflow = 0       : "
      f"{(zero_cashflow == 0).sum():,} "
      f"({(zero_cashflow == 0).mean()*100:.1f}%)")
print(f"  Cashflow < 0       : "
      f"{(zero_cashflow < 0).sum():,} "
      f"({(zero_cashflow < 0).mean()*100:.1f}%)")
print(f"  Cashflow > 0       : "
      f"{(zero_cashflow > 0).sum():,} "
      f"({(zero_cashflow > 0).mean()*100:.1f}%)")
print(f"  Mean Cashflow      : "
      f"R{zero_cashflow.mean():,.2f}")

# ── Summary conclusion ────────────────────────────────────
print(f"\n{'=' * 60}")
print("INTERPRETATION SUMMARY")
print("=" * 60)
print(f"""
  If zero-day records show:
  - Overwhelmingly Normal debtor status
  - Total Outstanding = 0 for most records
  - Negative cashflow (payment received)
  → Supports Explanation 1: Genuine same-day resolution

  If zero-day records show:
  - Mix of problematic debtor statuses
  - Positive outstanding balances
  - Zero or missing cashflow
  → Supports Explanation 2: Incomplete system recording

  This finding must be documented as a data limitation
  in the report regardless of which explanation dominates.
""")

ZERO vs NON-ZERO PROCESSING TIME — OVERVIEW

  Zero-day records     : 28,041 (56.1%)
  Non-zero-day records : 21,959 (43.9%)
  Total records        : 50,000

DEBTOR STATUS — ZERO-DAY vs NON-ZERO-DAY

  Debtor Status                         Zero-day    Zero%    NonZero   NonZero%
  ---------------------------------------------------------------------------
  Normal                                  26,750    95.4%     20,370      92.8%
  Payment arrangement                        367     1.3%        478       2.2%
  Deceased                                   287     1.0%        302       1.4%
  Handed over ACK                            206     0.7%        158       0.7%
  To be handed over                          171     0.6%        209       1.0%
  Handed over OTHER                           79     0.3%         23       0.1%
  WCA                                         49     0.2%         83       0.4%
  private patient                             40     0.1%         41       0.2%
  